File: python/vibedds/wire.py
"""RTPS message builder and parser.

Converts between bytes on the wire and RtpsMessage/submessage objects.
"""

from __future__ import annotations

import struct
from typing import TYPE_CHECKING

from vibedds.constants import (
    RTPS_MAGIC, RTPS_VERSION_MAJOR, RTPS_VERSION_MINOR, VENDOR_ID,
    SUBMSG_PAD, SUBMSG_ACKNACK, SUBMSG_HEARTBEAT, SUBMSG_GAP,
    SUBMSG_INFO_TS, SUBMSG_INFO_SRC, SUBMSG_INFO_DST, SUBMSG_DATA,
    FLAG_ENDIAN, FLAG_DATA_Q, FLAG_DATA_D, FLAG_DATA_K,
)
from vibedds.types import (
    GuidPrefix, EntityId, SequenceNumber, SequenceNumberSet,
    Timestamp, ProtocolVersion, VendorId,
)
from vibedds.messages import (
    RtpsHeader, RtpsMessage,
    DataSubmessage, HeartbeatSubmessage, AckNackSubmessage, GapSubmessage,
    InfoTimestampSubmessage, InfoDestinationSubmessage, InfoSourceSubmessage,
    PadSubmessage, Submessage,
)
from vibedds.cdr import CdrSerializer, CdrDeserializer


class RtpsMessageBuilder:
    """Builds an RTPS message from header + submessages into wire bytes."""

    def __init__(self, guid_prefix: GuidPrefix):
        self._guid_prefix = guid_prefix
        self._submessages: list[bytes] = []

    def _submessage_header(self, submsg_id: int, flags: int, data: bytes) -> bytes:
        """Build a 4-byte submessage header + data."""
        # Header: submessageId(1) + flags(1) + octetsToNextHeader(2, LE always)
        length = len(data)
        header = struct.pack("<BBH", submsg_id, flags, length)
        return header + data

    def add_info_ts(self, timestamp: Timestamp | None = None) -> None:
        """Add INFO_TS submessage."""
        flags = FLAG_ENDIAN  # little-endian
        if timestamp is None:
            flags |= 0x02  # invalidate flag
            data = b''
        else:
            data = timestamp.to_bytes("<")
        self._submessages.append(
            self._submessage_header(SUBMSG_INFO_TS, flags, data)
        )

    def add_info_dst(self, guid_prefix: GuidPrefix) -> None:
        """Add INFO_DST submessage."""
        flags = FLAG_ENDIAN
        self._submessages.append(
            self._submessage_header(SUBMSG_INFO_DST, flags, guid_prefix.to_bytes())
        )

    def add_data(
        self,
        reader_id: EntityId,
        writer_id: EntityId,
        writer_sn: SequenceNumber,
        serialized_payload: bytes | None = None,
        inline_qos: bytes | None = None,
        key_payload: bool = False,
    ) -> None:
        """Add DATA submessage."""
        flags = FLAG_ENDIAN
        if inline_qos is not None:
            flags |= FLAG_DATA_Q
        if serialized_payload is not None:
            if key_payload:
                flags |= FLAG_DATA_K
            else:
                flags |= FLAG_DATA_D

        buf = bytearray()
        # extraFlags (2 bytes, always 0)
        buf.extend(struct.pack("<H", 0))
        # octetsToInlineQos (2 bytes, always 16 from start of readerEntityId)
        buf.extend(struct.pack("<H", 16))
        # readerEntityId (4 bytes)
        buf.extend(reader_id.to_bytes())
        # writerEntityId (4 bytes)
        buf.extend(writer_id.to_bytes())
        # writerSeqNumber (8 bytes)
        buf.extend(struct.pack("<iI", writer_sn.high, writer_sn.low))
        # inline QoS
        if inline_qos is not None:
            buf.extend(inline_qos)
        # serialized payload
        if serialized_payload is not None:
            buf.extend(serialized_payload)

        self._submessages.append(
            self._submessage_header(SUBMSG_DATA, flags, bytes(buf))
        )

    def add_heartbeat(
        self,
        reader_id: EntityId,
        writer_id: EntityId,
        first_sn: SequenceNumber,
        last_sn: SequenceNumber,
        count: int,
        final: bool = False,
        liveliness: bool = False,
    ) -> None:
        """Add HEARTBEAT submessage."""
        flags = FLAG_ENDIAN
        if final:
            flags |= 0x02  # F
        if liveliness:
            flags |= 0x04  # L

        buf = bytearray()
        buf.extend(reader_id.to_bytes())
        buf.extend(writer_id.to_bytes())
        buf.extend(struct.pack("<iI", first_sn.high, first_sn.low))
        buf.extend(struct.pack("<iI", last_sn.high, last_sn.low))
        buf.extend(struct.pack("<I", count))

        self._submessages.append(
            self._submessage_header(SUBMSG_HEARTBEAT, flags, bytes(buf))
        )

    def add_acknack(
        self,
        reader_id: EntityId,
        writer_id: EntityId,
        reader_sn_state: SequenceNumberSet,
        count: int,
        final: bool = False,
    ) -> None:
        """Add ACKNACK submessage."""
        flags = FLAG_ENDIAN
        if final:
            flags |= 0x02

        ser = CdrSerializer("<")
        ser.write_bytes_raw(reader_id.to_bytes())
        ser.write_bytes_raw(writer_id.to_bytes())
        reader_sn_state.serialize(ser)
        ser.write_uint32(count)

        self._submessages.append(
            self._submessage_header(SUBMSG_ACKNACK, flags, ser.getvalue())
        )

    def add_gap(
        self,
        reader_id: EntityId,
        writer_id: EntityId,
        gap_start: SequenceNumber,
        gap_list: SequenceNumberSet,
    ) -> None:
        """Add GAP submessage."""
        flags = FLAG_ENDIAN

        ser = CdrSerializer("<")
        ser.write_bytes_raw(reader_id.to_bytes())
        ser.write_bytes_raw(writer_id.to_bytes())
        gap_start.serialize(ser)
        gap_list.serialize(ser)

        self._submessages.append(
            self._submessage_header(SUBMSG_GAP, flags, ser.getvalue())
        )

    def build(self) -> bytes:
        """Build the complete RTPS message bytes."""
        buf = bytearray()
        # RTPS header (20 bytes)
        buf.extend(RTPS_MAGIC)
        buf.extend(bytes([RTPS_VERSION_MAJOR, RTPS_VERSION_MINOR]))
        buf.extend(bytes(VENDOR_ID))
        buf.extend(self._guid_prefix.to_bytes())
        # Submessages
        for sm in self._submessages:
            buf.extend(sm)
        return bytes(buf)


class RtpsMessageParser:
    """Parses RTPS wire bytes into RtpsMessage objects."""

    @staticmethod
    def parse(data: bytes | bytearray | memoryview) -> RtpsMessage:
        """Parse a complete RTPS message from wire bytes.

        Raises ValueError on invalid data.
        """
        data = bytes(data)
        if len(data) < 20:
            raise ValueError(f"RTPS message too short: {len(data)} bytes (need >= 20)")

        # Validate magic
        if data[0:4] != RTPS_MAGIC:
            raise ValueError(f"Invalid RTPS magic: {data[0:4]!r}")

        # Parse header
        version = ProtocolVersion(major=data[4], minor=data[5])
        vendor_id = VendorId(vendor=(data[6], data[7]))
        guid_prefix = GuidPrefix(data[8:20])
        header = RtpsHeader(version=version, vendor_id=vendor_id, guid_prefix=guid_prefix)

        # Parse submessages
        submessages: list[Submessage] = []
        pos = 20
        while pos < len(data):
            if pos + 4 > len(data):
                break  # incomplete submessage header

            submsg_id = data[pos]
            flags = data[pos + 1]
            octets_to_next = struct.unpack_from("<H", data, pos + 2)[0]
            pos += 4

            little_endian = bool(flags & FLAG_ENDIAN)
            endian = "<" if little_endian else ">"

            # Get submessage body
            if octets_to_next == 0 and pos < len(data):
                # Last submessage may use 0 to mean "rest of message"
                body = data[pos:]
                pos = len(data)
            else:
                body = data[pos:pos + octets_to_next]
                pos += octets_to_next

            submsg = RtpsMessageParser._parse_submessage(submsg_id, flags, endian, body)
            if submsg is not None:
                submessages.append(submsg)

        return RtpsMessage(header=header, submessages=submessages)

    @staticmethod
    def _parse_submessage(
        submsg_id: int, flags: int, endian: str, body: bytes
    ) -> Submessage | None:
        """Parse a single submessage body. Returns None for unknown types."""
        try:
            if submsg_id == SUBMSG_DATA:
                return RtpsMessageParser._parse_data(flags, endian, body)
            elif submsg_id == SUBMSG_HEARTBEAT:
                return RtpsMessageParser._parse_heartbeat(flags, endian, body)
            elif submsg_id == SUBMSG_ACKNACK:
                return RtpsMessageParser._parse_acknack(flags, endian, body)
            elif submsg_id == SUBMSG_GAP:
                return RtpsMessageParser._parse_gap(flags, endian, body)
            elif submsg_id == SUBMSG_INFO_TS:
                return RtpsMessageParser._parse_info_ts(flags, endian, body)
            elif submsg_id == SUBMSG_INFO_DST:
                return RtpsMessageParser._parse_info_dst(flags, endian, body)
            elif submsg_id == SUBMSG_INFO_SRC:
                return RtpsMessageParser._parse_info_src(flags, endian, body)
            elif submsg_id == SUBMSG_PAD:
                return PadSubmessage(flags=flags)
            else:
                # Unknown submessage — skip
                return None
        except Exception:
            # Malformed submessage — skip rather than crash
            return None

    @staticmethod
    def _parse_data(flags: int, endian: str, body: bytes) -> DataSubmessage:
        has_inline_qos = bool(flags & FLAG_DATA_Q)
        has_data = bool(flags & FLAG_DATA_D)
        has_key = bool(flags & FLAG_DATA_K)

        # extraFlags(2) + octetsToInlineQos(2) + readerId(4) + writerId(4) + writerSN(8) = 20
        if len(body) < 20:
            raise ValueError("DATA submessage too short")

        octets_to_inline_qos = struct.unpack_from(endian + "H", body, 2)[0]
        reader_id = EntityId(body[4:8])
        writer_id = EntityId(body[8:12])

        sn_high = struct.unpack_from(endian + "i", body, 12)[0]
        sn_low = struct.unpack_from(endian + "I", body, 16)[0]
        writer_sn = SequenceNumber(high=sn_high, low=sn_low)

        payload_offset = 4 + octets_to_inline_qos  # from start of body
        inline_qos = None
        serialized_payload = None

        if has_inline_qos:
            # Find PID_SENTINEL to determine end of inline QoS
            qos_start = payload_offset
            qos_end = RtpsMessageParser._find_sentinel(body, qos_start, endian)
            inline_qos = body[qos_start:qos_end]
            payload_offset = qos_end

        if has_data or has_key:
            serialized_payload = body[payload_offset:]

        return DataSubmessage(
            flags=flags,
            reader_id=reader_id,
            writer_id=writer_id,
            writer_sn=writer_sn,
            inline_qos=inline_qos,
            serialized_payload=serialized_payload,
        )

    @staticmethod
    def _find_sentinel(data: bytes, start: int, endian: str) -> int:
        """Scan parameter list for PID_SENTINEL, return position after sentinel."""
        pos = start
        while pos + 4 <= len(data):
            pid = struct.unpack_from(endian + "H", data, pos)[0]
            length = struct.unpack_from(endian + "H", data, pos + 2)[0]
            pos += 4 + length
            if pid == 0x0001:  # PID_SENTINEL
                return pos
        return pos

    @staticmethod
    def _parse_heartbeat(flags: int, endian: str, body: bytes) -> HeartbeatSubmessage:
        if len(body) < 28:
            raise ValueError("HEARTBEAT submessage too short")

        reader_id = EntityId(body[0:4])
        writer_id = EntityId(body[4:8])
        first_high = struct.unpack_from(endian + "i", body, 8)[0]
        first_low = struct.unpack_from(endian + "I", body, 12)[0]
        last_high = struct.unpack_from(endian + "i", body, 16)[0]
        last_low = struct.unpack_from(endian + "I", body, 20)[0]
        count = struct.unpack_from(endian + "I", body, 24)[0]

        return HeartbeatSubmessage(
            flags=flags,
            reader_id=reader_id,
            writer_id=writer_id,
            first_sn=SequenceNumber(high=first_high, low=first_low),
            last_sn=SequenceNumber(high=last_high, low=last_low),
            count=count,
        )

    @staticmethod
    def _parse_acknack(flags: int, endian: str, body: bytes) -> AckNackSubmessage:
        reader_id = EntityId(body[0:4])
        writer_id = EntityId(body[4:8])

        de = CdrDeserializer(body[8:], endian)
        sn_state = SequenceNumberSet.deserialize(de)
        count = de.read_uint32()

        return AckNackSubmessage(
            flags=flags,
            reader_id=reader_id,
            writer_id=writer_id,
            reader_sn_state=sn_state,
            count=count,
        )

    @staticmethod
    def _parse_gap(flags: int, endian: str, body: bytes) -> GapSubmessage:
        reader_id = EntityId(body[0:4])
        writer_id = EntityId(body[4:8])

        de = CdrDeserializer(body[8:], endian)
        gap_start = SequenceNumber.deserialize(de)
        gap_list = SequenceNumberSet.deserialize(de)

        return GapSubmessage(
            flags=flags,
            reader_id=reader_id,
            writer_id=writer_id,
            gap_start=gap_start,
            gap_list=gap_list,
        )

    @staticmethod
    def _parse_info_ts(flags: int, endian: str, body: bytes) -> InfoTimestampSubmessage:
        invalidate = bool(flags & 0x02)
        if invalidate or len(body) < 8:
            return InfoTimestampSubmessage(flags=flags, timestamp=None)
        ts = Timestamp.from_bytes(body[:8], endian)
        return InfoTimestampSubmessage(flags=flags, timestamp=ts)

    @staticmethod
    def _parse_info_dst(flags: int, endian: str, body: bytes) -> InfoDestinationSubmessage:
        if len(body) < 12:
            raise ValueError("INFO_DST too short")
        return InfoDestinationSubmessage(
            flags=flags,
            guid_prefix=GuidPrefix(body[:12]),
        )

    @staticmethod
    def _parse_info_src(flags: int, endian: str, body: bytes) -> InfoSourceSubmessage:
        if len(body) < 20:
            raise ValueError("INFO_SRC too short")
        # unused(4) + version(2) + vendorId(2) + guidPrefix(12) = 20
        version = ProtocolVersion(major=body[4], minor=body[5])
        vendor_id = VendorId(vendor=(body[6], body[7]))
        guid_prefix = GuidPrefix(body[8:20])
        return InfoSourceSubmessage(
            flags=flags,
            protocol_version=version,
            vendor_id=vendor_id,
            guid_prefix=guid_prefix,
        )

File: rust/vibedds/src/wire.rs
/// RTPS message builder and parser.

use std::io;

use crate::cdr::{CdrDeserializer, Endian};
use crate::constants::*;
use crate::messages::*;
use crate::types::*;

/// Builds an RTPS message from header + submessages.
pub struct RtpsMessageBuilder {
    guid_prefix: GuidPrefix,
    submessages: Vec<Vec<u8>>,
}

impl RtpsMessageBuilder {
    pub fn new(guid_prefix: GuidPrefix) -> Self {
        Self {
            guid_prefix,
            submessages: Vec::new(),
        }
    }

    fn submessage_bytes(submsg_id: u8, flags: u8, data: &[u8]) -> Vec<u8> {
        let mut buf = Vec::with_capacity(4 + data.len());
        buf.push(submsg_id);
        buf.push(flags);
        buf.extend_from_slice(&(data.len() as u16).to_le_bytes());
        buf.extend_from_slice(data);
        buf
    }

    pub fn add_info_ts(&mut self, timestamp: Option<Timestamp>) {
        let mut flags = FLAG_ENDIAN;
        let data = match timestamp {
            Some(ts) => ts.to_bytes_le().to_vec(),
            None => {
                flags |= 0x02;
                Vec::new()
            }
        };
        self.submessages
            .push(Self::submessage_bytes(SUBMSG_INFO_TS, flags, &data));
    }

    pub fn add_info_dst(&mut self, guid_prefix: GuidPrefix) {
        self.submessages.push(Self::submessage_bytes(
            SUBMSG_INFO_DST,
            FLAG_ENDIAN,
            &guid_prefix.0,
        ));
    }

    pub fn add_data(
        &mut self,
        reader_id: EntityId,
        writer_id: EntityId,
        writer_sn: SequenceNumber,
        serialized_payload: Option<&[u8]>,
        inline_qos: Option<&[u8]>,
        key_payload: bool,
    ) {
        let mut flags = FLAG_ENDIAN;
        if inline_qos.is_some() {
            flags |= FLAG_DATA_Q;
        }
        if let Some(_) = serialized_payload {
            if key_payload {
                flags |= FLAG_DATA_K;
            } else {
                flags |= FLAG_DATA_D;
            }
        }

        let mut buf = Vec::new();
        // extraFlags
        buf.extend_from_slice(&0u16.to_le_bytes());
        // octetsToInlineQos = 16
        buf.extend_from_slice(&16u16.to_le_bytes());
        // readerEntityId
        buf.extend_from_slice(&reader_id.0);
        // writerEntityId
        buf.extend_from_slice(&writer_id.0);
        // writerSeqNumber
        buf.extend_from_slice(&writer_sn.high.to_le_bytes());
        buf.extend_from_slice(&writer_sn.low.to_le_bytes());
        // inline QoS
        if let Some(qos) = inline_qos {
            buf.extend_from_slice(qos);
        }
        // payload
        if let Some(payload) = serialized_payload {
            buf.extend_from_slice(payload);
        }

        self.submessages
            .push(Self::submessage_bytes(SUBMSG_DATA, flags, &buf));
    }

    pub fn add_heartbeat(
        &mut self,
        reader_id: EntityId,
        writer_id: EntityId,
        first_sn: SequenceNumber,
        last_sn: SequenceNumber,
        count: u32,
        final_flag: bool,
        liveliness: bool,
    ) {
        let mut flags = FLAG_ENDIAN;
        if final_flag {
            flags |= 0x02;
        }
        if liveliness {
            flags |= 0x04;
        }

        let mut buf = Vec::with_capacity(28);
        buf.extend_from_slice(&reader_id.0);
        buf.extend_from_slice(&writer_id.0);
        buf.extend_from_slice(&first_sn.high.to_le_bytes());
        buf.extend_from_slice(&first_sn.low.to_le_bytes());
        buf.extend_from_slice(&last_sn.high.to_le_bytes());
        buf.extend_from_slice(&last_sn.low.to_le_bytes());
        buf.extend_from_slice(&count.to_le_bytes());

        self.submessages
            .push(Self::submessage_bytes(SUBMSG_HEARTBEAT, flags, &buf));
    }

    pub fn build(&self) -> Vec<u8> {
        let mut buf = Vec::new();
        // RTPS header (20 bytes)
        buf.extend_from_slice(&RTPS_MAGIC);
        buf.push(RTPS_VERSION_MAJOR);
        buf.push(RTPS_VERSION_MINOR);
        buf.extend_from_slice(&VENDOR_ID);
        buf.extend_from_slice(&self.guid_prefix.0);
        // Submessages
        for sm in &self.submessages {
            buf.extend_from_slice(sm);
        }
        buf
    }
}

/// Parses RTPS wire bytes into RtpsMessage.
pub fn parse_rtps_message(data: &[u8]) -> io::Result<RtpsMessage> {
    if data.len() < 20 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("RTPS message too short: {} bytes", data.len()),
        ));
    }
    if &data[0..4] != &RTPS_MAGIC {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "Invalid RTPS magic",
        ));
    }

    let version = ProtocolVersion::new(data[4], data[5]);
    let vendor_id = VendorId([data[6], data[7]]);
    let mut prefix = [0u8; 12];
    prefix.copy_from_slice(&data[8..20]);
    let guid_prefix = GuidPrefix(prefix);

    let header = RtpsHeader {
        version,
        vendor_id,
        guid_prefix,
    };

    let mut submessages = Vec::new();
    let mut pos = 20;

    while pos + 4 <= data.len() {
        let submsg_id = data[pos];
        let flags = data[pos + 1];
        let octets_to_next = u16::from_le_bytes([data[pos + 2], data[pos + 3]]) as usize;
        pos += 4;

        let little_endian = (flags & FLAG_ENDIAN) != 0;
        let endian = if little_endian {
            Endian::Little
        } else {
            Endian::Big
        };

        let body_end = if octets_to_next == 0 && pos < data.len() {
            data.len()
        } else {
            (pos + octets_to_next).min(data.len())
        };
        let body = &data[pos..body_end];
        pos = body_end;

        if let Some(sm) = parse_submessage(submsg_id, flags, endian, body) {
            submessages.push(sm);
        }
    }

    Ok(RtpsMessage {
        header,
        submessages,
    })
}

fn parse_submessage(submsg_id: u8, flags: u8, endian: Endian, body: &[u8]) -> Option<Submessage> {
    match submsg_id {
        SUBMSG_DATA => parse_data(flags, endian, body).map(Submessage::Data),
        SUBMSG_HEARTBEAT => parse_heartbeat(flags, endian, body).map(Submessage::Heartbeat),
        SUBMSG_ACKNACK => parse_acknack(flags, endian, body).map(Submessage::AckNack),
        SUBMSG_INFO_TS => parse_info_ts(flags, endian, body).map(Submessage::InfoTimestamp),
        SUBMSG_INFO_DST => parse_info_dst(flags, body).map(Submessage::InfoDestination),
        SUBMSG_PAD => Some(Submessage::Pad(PadSubmessage { flags })),
        _ => None,
    }
}

fn parse_data(flags: u8, endian: Endian, body: &[u8]) -> Option<DataSubmessage> {
    if body.len() < 20 {
        return None;
    }
    let has_data = (flags & FLAG_DATA_D) != 0;
    let has_key = (flags & FLAG_DATA_K) != 0;
    let has_qos = (flags & FLAG_DATA_Q) != 0;

    let octets_to_inline_qos = match endian {
        Endian::Little => u16::from_le_bytes([body[2], body[3]]) as usize,
        Endian::Big => u16::from_be_bytes([body[2], body[3]]) as usize,
    };

    let mut reader_id = [0u8; 4];
    reader_id.copy_from_slice(&body[4..8]);
    let mut writer_id = [0u8; 4];
    writer_id.copy_from_slice(&body[8..12]);

    let sn_high = match endian {
        Endian::Little => i32::from_le_bytes(body[12..16].try_into().unwrap()),
        Endian::Big => i32::from_be_bytes(body[12..16].try_into().unwrap()),
    };
    let sn_low = match endian {
        Endian::Little => u32::from_le_bytes(body[16..20].try_into().unwrap()),
        Endian::Big => u32::from_be_bytes(body[16..20].try_into().unwrap()),
    };

    let mut payload_offset = 4 + octets_to_inline_qos;
    let inline_qos = if has_qos && payload_offset < body.len() {
        let qos_start = payload_offset;
        let qos_end = find_sentinel(body, qos_start, endian);
        let qos = body[qos_start..qos_end].to_vec();
        payload_offset = qos_end;
        Some(qos)
    } else {
        None
    };

    let serialized_payload = if (has_data || has_key) && payload_offset < body.len() {
        Some(body[payload_offset..].to_vec())
    } else {
        None
    };

    Some(DataSubmessage {
        flags,
        reader_id: EntityId(reader_id),
        writer_id: EntityId(writer_id),
        writer_sn: SequenceNumber {
            high: sn_high,
            low: sn_low,
        },
        inline_qos,
        serialized_payload,
    })
}

fn find_sentinel(data: &[u8], start: usize, endian: Endian) -> usize {
    let mut pos = start;
    while pos + 4 <= data.len() {
        let pid = match endian {
            Endian::Little => u16::from_le_bytes([data[pos], data[pos + 1]]),
            Endian::Big => u16::from_be_bytes([data[pos], data[pos + 1]]),
        };
        let length = match endian {
            Endian::Little => u16::from_le_bytes([data[pos + 2], data[pos + 3]]) as usize,
            Endian::Big => u16::from_be_bytes([data[pos + 2], data[pos + 3]]) as usize,
        };
        pos += 4 + length;
        if pid == PID_SENTINEL {
            return pos;
        }
    }
    pos
}

fn parse_heartbeat(flags: u8, endian: Endian, body: &[u8]) -> Option<HeartbeatSubmessage> {
    if body.len() < 28 {
        return None;
    }
    let mut reader_id = [0u8; 4];
    reader_id.copy_from_slice(&body[0..4]);
    let mut writer_id = [0u8; 4];
    writer_id.copy_from_slice(&body[4..8]);

    macro_rules! read_i32 {
        ($offset:expr) => {
            match endian {
                Endian::Little => i32::from_le_bytes(body[$offset..$offset + 4].try_into().unwrap()),
                Endian::Big => i32::from_be_bytes(body[$offset..$offset + 4].try_into().unwrap()),
            }
        };
    }
    macro_rules! read_u32 {
        ($offset:expr) => {
            match endian {
                Endian::Little => u32::from_le_bytes(body[$offset..$offset + 4].try_into().unwrap()),
                Endian::Big => u32::from_be_bytes(body[$offset..$offset + 4].try_into().unwrap()),
            }
        };
    }

    Some(HeartbeatSubmessage {
        flags,
        reader_id: EntityId(reader_id),
        writer_id: EntityId(writer_id),
        first_sn: SequenceNumber {
            high: read_i32!(8),
            low: read_u32!(12),
        },
        last_sn: SequenceNumber {
            high: read_i32!(16),
            low: read_u32!(20),
        },
        count: read_u32!(24),
    })
}

fn parse_acknack(flags: u8, endian: Endian, body: &[u8]) -> Option<AckNackSubmessage> {
    if body.len() < 8 {
        return None;
    }
    let mut reader_id = [0u8; 4];
    reader_id.copy_from_slice(&body[0..4]);
    let mut writer_id = [0u8; 4];
    writer_id.copy_from_slice(&body[4..8]);

    let mut de = CdrDeserializer::new(&body[8..], endian);
    let sn_state = SequenceNumberSet::deserialize(&mut de).ok()?;
    let count = de.read_u32().ok()?;

    Some(AckNackSubmessage {
        flags,
        reader_id: EntityId(reader_id),
        writer_id: EntityId(writer_id),
        reader_sn_state: sn_state,
        count,
    })
}

fn parse_info_ts(flags: u8, endian: Endian, body: &[u8]) -> Option<InfoTimestampSubmessage> {
    let invalidate = (flags & 0x02) != 0;
    if invalidate || body.len() < 8 {
        return Some(InfoTimestampSubmessage {
            flags,
            timestamp: None,
        });
    }
    let ts = match endian {
        Endian::Little => Timestamp {
            seconds: i32::from_le_bytes(body[0..4].try_into().unwrap()),
            fraction: u32::from_le_bytes(body[4..8].try_into().unwrap()),
        },
        Endian::Big => Timestamp {
            seconds: i32::from_be_bytes(body[0..4].try_into().unwrap()),
            fraction: u32::from_be_bytes(body[4..8].try_into().unwrap()),
        },
    };
    Some(InfoTimestampSubmessage {
        flags,
        timestamp: Some(ts),
    })
}

fn parse_info_dst(flags: u8, body: &[u8]) -> Option<InfoDestinationSubmessage> {
    if body.len() < 12 {
        return None;
    }
    let mut prefix = [0u8; 12];
    prefix.copy_from_slice(&body[..12]);
    Some(InfoDestinationSubmessage {
        flags,
        guid_prefix: GuidPrefix(prefix),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_header_wire_format() {
        let prefix = GuidPrefix([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]);
        let builder = RtpsMessageBuilder::new(prefix);
        let data = builder.build();
        assert_eq!(data.len(), 20);
        assert_eq!(&data[0..4], b"RTPS");
        assert_eq!(data[4], RTPS_VERSION_MAJOR);
        assert_eq!(data[5], RTPS_VERSION_MINOR);
        assert_eq!(data[6], 0xFF);
        assert_eq!(data[7], 0x01);
        assert_eq!(&data[8..20], &prefix.0);
    }

    #[test]
    fn test_parse_header() {
        let prefix = GuidPrefix([0xAA; 12]);
        let builder = RtpsMessageBuilder::new(prefix);
        let msg = parse_rtps_message(&builder.build()).unwrap();
        assert_eq!(msg.header.version, ProtocolVersion::new(2, 5));
        assert_eq!(msg.header.vendor_id, VendorId([0xFF, 0x01]));
        assert_eq!(msg.header.guid_prefix, prefix);
    }

    #[test]
    fn test_info_ts_roundtrip() {
        let prefix = GuidPrefix([1; 12]);
        let ts = Timestamp {
            seconds: 1000,
            fraction: 500,
        };
        let mut builder = RtpsMessageBuilder::new(prefix);
        builder.add_info_ts(Some(ts));
        let msg = parse_rtps_message(&builder.build()).unwrap();
        assert_eq!(msg.submessages.len(), 1);
        if let Submessage::InfoTimestamp(its) = &msg.submessages[0] {
            assert_eq!(its.timestamp, Some(ts));
        } else {
            panic!("Expected InfoTimestamp");
        }
    }

    #[test]
    fn test_data_roundtrip() {
        let prefix = GuidPrefix([1; 12]);
        let reader_id = EntityId(ENTITYID_SPDP_BUILTIN_PARTICIPANT_READER);
        let writer_id = EntityId(ENTITYID_SPDP_BUILTIN_PARTICIPANT_WRITER);
        let sn = SequenceNumber::from_value(1);
        let payload = b"\xDE\xAD\xBE\xEF";

        let mut builder = RtpsMessageBuilder::new(prefix);
        builder.add_data(reader_id, writer_id, sn, Some(payload), None, false);
        let msg = parse_rtps_message(&builder.build()).unwrap();
        assert_eq!(msg.submessages.len(), 1);
        if let Submessage::Data(d) = &msg.submessages[0] {
            assert_eq!(d.reader_id, reader_id);
            assert_eq!(d.writer_id, writer_id);
            assert_eq!(d.writer_sn, sn);
            assert_eq!(d.serialized_payload.as_deref(), Some(payload.as_slice()));
        } else {
            panic!("Expected Data");
        }
    }

    #[test]
    fn test_heartbeat_roundtrip() {
        let prefix = GuidPrefix([1; 12]);
        let reader_id = EntityId(ENTITYID_SPDP_BUILTIN_PARTICIPANT_READER);
        let writer_id = EntityId(ENTITYID_SPDP_BUILTIN_PARTICIPANT_WRITER);

        let mut builder = RtpsMessageBuilder::new(prefix);
        builder.add_heartbeat(
            reader_id,
            writer_id,
            SequenceNumber::from_value(1),
            SequenceNumber::from_value(10),
            5,
            true,
            false,
        );
        let msg = parse_rtps_message(&builder.build()).unwrap();
        if let Submessage::Heartbeat(hb) = &msg.submessages[0] {
            assert_eq!(hb.first_sn.value(), 1);
            assert_eq!(hb.last_sn.value(), 10);
            assert_eq!(hb.count, 5);
        } else {
            panic!("Expected Heartbeat");
        }
    }

    #[test]
    fn test_multi_submessage() {
        let prefix = GuidPrefix([1; 12]);
        let dst = GuidPrefix([2; 12]);
        let ts = Timestamp {
            seconds: 12345,
            fraction: 0,
        };
        let reader_id = EntityId(ENTITYID_SPDP_BUILTIN_PARTICIPANT_READER);
        let writer_id = EntityId(ENTITYID_SPDP_BUILTIN_PARTICIPANT_WRITER);

        let mut builder = RtpsMessageBuilder::new(prefix);
        builder.add_info_ts(Some(ts));
        builder.add_info_dst(dst);
        builder.add_data(
            reader_id,
            writer_id,
            SequenceNumber::from_value(1),
            Some(b"\xAB\xCD"),
            None,
            false,
        );
        builder.add_heartbeat(
            reader_id,
            writer_id,
            SequenceNumber::from_value(1),
            SequenceNumber::from_value(1),
            1,
            false,
            false,
        );

        let msg = parse_rtps_message(&builder.build()).unwrap();
        assert_eq!(msg.submessages.len(), 4);
        assert!(matches!(&msg.submessages[0], Submessage::InfoTimestamp(_)));
        assert!(matches!(&msg.submessages[1], Submessage::InfoDestination(_)));
        assert!(matches!(&msg.submessages[2], Submessage::Data(_)));
        assert!(matches!(&msg.submessages[3], Submessage::Heartbeat(_)));
    }

    #[test]
    fn test_parse_too_short() {
        assert!(parse_rtps_message(&[0; 10]).is_err());
    }

    #[test]
    fn test_parse_bad_magic() {
        let mut data = [0u8; 20];
        data[0..4].copy_from_slice(b"XXXX");
        assert!(parse_rtps_message(&data).is_err());
    }
}
